File: Src/SerialCom.c
/**
 * \file
 * \author	<a href="http://www.innomatic.ca">innomatic</a>
 * \copyright <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>.
 */
#include <stdbool.h>
#include "board.h"
#include "ByteQueue.h"
#include "SerialCom.h"

// packet decoding state machine states
#define PKT_STATE_HDR			0
#define PKT_STATE_LEN			1
#define PKT_STATE_PLD			2
#define PKT_STATE_CSM			3

#ifndef CLEAR_UART_RX_IT
	#error CLEAR_UART_RX_IT not defined
#endif

/**
 * This function runs packet decoding state machine. It takes stream of serial
 * data one byte at a time and returns the current state of the state machine.
 * Upon successful retrieval of a full packet, it fills the array buffer[]
 * with the received payload.
 *
 * To use this function, define an uint8_t array for the payload data storage
 * and call this function whenever new byte is arrived. The best place to put
 * this function is inside the UART check routine (not inside the interrupt
 * service callback function).
 *
 * \param	byte received byte
 * \param   buffer packet will be retured here
 * \return  pkt_status status of the state machine
 *
\code
uint8_t buffer[MAX_PKTSIZE];
pkt_status pstatus;

if(ByteQueue_Get(&new_byate)
{
    // new data byte is in new_byte
    if(PKT_RECEIVED == SerialCom_Decoder(new_byte, buffer))
    {
        // packet data is in the buffer
    }
}
\endcode
 */
pkt_status SerialCom_Decoder(uint8_t byte, uint8_t *buffer)
{
	static uint8_t state = PKT_STATE_HDR;
	static uint8_t index = 0;
	static uint8_t packet[MAX_PKTSIZE] = {0};
	static uint8_t csum = 0;

	int i;

	// waiting for the header byte
	if(state == PKT_STATE_HDR)
	{
		if(byte == PKT_HEADR)
		{
			// store the header byte
			packet[0] = byte;
			// proceed to the next state
			state = PKT_STATE_LEN;
		}
		else if(byte == PKT_ACK)
		{
			// ACK received but do not change the state
			return ACK_RECEIVED;
		}
		else if(byte == PKT_NAK)
		{
			// NAK received but do not change the state
			return NAK_RECEIVED;
		}
		else if(byte == PKT_IAM)
		{
			// IAM received but do not change the state
			return IAM_RECEIVED;
		}
	}
	// waiting for the length byte
	else if(state == PKT_STATE_LEN)
	{
		// store the length byte
		packet[1] = byte;

		// invalid payload size
		if(byte > MAX_PAYLOAD)
		{
			// start all over
			state = PKT_STATE_HDR;
			// report size error
			return PKT_SIZE_ERR;
		}
		// length byte is valid
		else
		{
			// reset index
			index = 2;
			// clear chesum byte
			csum = 0;
			// proceed to the next state
			state = PKT_STATE_PLD;
		}
	}
	// waiting for the payload
	else if(state == PKT_STATE_PLD)
	{
		// collect data
		packet[index++] = byte;
		// process checksum
		csum ^= byte;
		// proceed to the next if all payload is collected
		if(index == (packet[1] + 2))
		{
			state = PKT_STATE_CSM;
		}
	}
	// waiting for the checksum byte
	else if(state == PKT_STATE_CSM)
	{
		// collect data
		packet[index] = byte;
		// checksum matches
		if(byte == csum)
		{
			// copy packet to the buffer
			for(i = 0; i < packet[1] + 3; i++)
			{
				buffer[i] = packet[i];
			}
			// start all over again
			state = PKT_STATE_HDR;
			// valid packet arrived
			return PKT_RECEIVED;
		}
		// checksum does not match
		else
		{
			// start all over
			state = PKT_STATE_HDR;
			// checksum error
			return PKT_CSUM_ERR;
		}
	}

	return PKT_INPROCES;
}

/** Construct a packet from a give payload and send it via UART
 *
 *	\param payload payload data
 *	\param size size of the payload
 *
\code
uint8_t payload[5];

// pack signed int32 value in big endian order
payload[0] = RPT_S32XXX;
payload[1] = (value >> 24) & 0xff;
payload[2] = (value >> 16) & 0xff;
payload[3] = (value >> 8) & 0xff;
payload[4] = (value) & 0xff;

// send the packet
SerialCom_SendPacket(payload, 5);
\endcode
 */
void SerialCom_SendPacket(uint8_t *payload, int size)
{
	uint8_t csum = 0;
	uint8_t packet[MAX_PKTSIZE];
	int i;

	// header
	packet[0] = PKT_HEADR;
	// payload size
	packet[1] = size;
	// data
	for(i = 0; i < size; i++)
	{
		csum ^= payload[i];
		packet[2+i] = payload[i];
	}
	// checksum
	packet[2 + size] = csum;

	// send the array of bytes
	SerialCom_SendByteArray(packet, size+3);
}

/**	Put this inside of RX callback of the UART.
 *	
 *	\param byte data byte read from UART
 */
inline void SerialCom_RxRoutine(uint8_t byte)
{
	ByteQueue_Put(byte);
	CLEAR_UART_RX_IT;
}


#if UNIT_TEST
/**
 */
#define TO_HEX(a)		(((a) > 9) ? ((a)-10+'A'):((a)+'0'))

/** Call this function inside of main loop. It sends ASCII string via UART,
 *	which represents the state of the decoder in the following way.
 *
 *  xx.xx....xx (P) : packet received without error
 *  (E) : packet received with error
 *  (A) : ACK received
 *  (N) : NAK received
 *  (I) : IAM received
 */
void SerialCom_RcvTest(void)
{
	uint8_t u8val;
	uint8_t rx_buffer[MAX_PKTSIZE];
	static int index = 0;
	static uint8_t dbg_out[200];

	if(ByteQueue_Get(&u8val))
	{
		switch(SerialCom_Decoder(u8val, rx_buffer))
		{
		case PKT_INPROCES:
			dbg_out[index++] = TO_HEX(u8val>>4);
			dbg_out[index++] = TO_HEX(u8val&0x0f);
			dbg_out[index++] = '.';
			break;

		case PKT_RECEIVED:
			dbg_out[index++] = TO_HEX(u8val>>4);
			dbg_out[index++] = TO_HEX(u8val&0x0f);
			dbg_out[index++] = '(';
			dbg_out[index++] = 'P';
			dbg_out[index++] = ')';
			dbg_out[index++] = '\n';
			SerialCom_SendByteArray(dbg_out, index);
			index = 0;
			break;

		case ACK_RECEIVED:
			dbg_out[index++] = TO_HEX(u8val>>4);
			dbg_out[index++] = TO_HEX(u8val&0x0f);
			dbg_out[index++] = '(';
			dbg_out[index++] = 'A';
			dbg_out[index++] = ')';
			dbg_out[index++] = '\n';
			SerialCom_SendByteArray(dbg_out, index);
			index = 0;
			break;

		case NAK_RECEIVED:
			dbg_out[index++] = TO_HEX(u8val>>4);
			dbg_out[index++] = TO_HEX(u8val&0x0f);
			dbg_out[index++] = '(';
			dbg_out[index++] = 'N';
			dbg_out[index++] = ')';
			dbg_out[index++] = '\n';
			SerialCom_SendByteArray(dbg_out, index);
			index = 0;
			break;

		case IAM_RECEIVED:
			dbg_out[index++] = TO_HEX(u8val>>4);
			dbg_out[index++] = TO_HEX(u8val&0x0f);
			dbg_out[index++] = '(';
			dbg_out[index++] = 'I';
			dbg_out[index++] = ')';
			dbg_out[index++] = '\n';
			SerialCom_SendByteArray(dbg_out, index);
			index = 0;
			break;

		case PKT_SIZE_ERR:
		case PKT_CSUM_ERR:
			dbg_out[index++] = TO_HEX(u8val>>4);
			dbg_out[index++] = TO_HEX(u8val&0x0f);
			dbg_out[index++] = '(';
			dbg_out[index++] = 'E';
			dbg_out[index++] = ')';
			dbg_out[index++] = '\n';
			SerialCom_SendByteArray(dbg_out, index);
			index = 0;
			break;

		default:
			break;
		}
	}
}

#else

void SerialCom_RecvTest(void)
{
}

#endif

File: Python/SerialCom.py
#!/usr/bin/env python3

################################################################################
#
#   \file
#   \author     <a href="http://www.innomatic.ca">innomatic</a>
#   \brief      Python Proxy of SerialCom.
#   \copyright  <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>.
#
#   Definitions here must match with that of SerialComm.h
# 
#   \verbatim
#   Packet Structure
#
#   (HDR) (LEN) (PLD) (CSM)
#
#   (HDR) header single byte
#   (LEN) size of the payload, single byte
#   (PLD) payload of variable size
#   (CSM) checksum, single byte
#
#   Payload Structure
#
#   (CMD) (DAT)
#
#   (CMD) single byte command code (DAT) data of variable length
#   @endverbatim
#
#   Receiver is expected to respond with either single byte of ACK or NAK
#

MAX_PAYLOAD = 10
MAX_PACKET = MAX_PAYLOAD + 3

# header signature, ACK, NAK, ...
PKT_HEADR  = 0xf5
PKT_ACK    = 0xf6
PKT_NAK    = 0xf7
PKT_IAM    = 0xf8

# System control
SYS_SRESET = 0x00   # perform software reset
SYS_WRESET = 0x01   # perform watchdog reset
SYS_STPMOD = 0x02   # enter stop mode
SYS_DSLMOD = 0x03   # enter deep sleep mode
SYS_SLPMOD = 0x04   # enter sleep mode

# DIO control
DIO_SETVAL = 0x10   # DIO set value
DIO_RSTVAL = 0x11   # DIO reset value
DIO_GETVAL = 0x12   # DIO get value
DIO_TGLVAL = 0x13   # DIO toggle value

# ADC control
ADC_SETMAG = 0x20   # ADC set magnitude
ADC_SETFRQ = 0x21   # ADC set frequency
ADC_CONSTV = 0x23   # ADC dc output
ADC_SINEWV = 0x24   # ADC sine waveform
ADC_SWTHWV = 0x25   # ADC sawtooth wavefrom
ADC_TRNGWV = 0x26   # ADC triangle waveform

# DAC control
DAC_SETFRQ = 0x30   # DAC set sample frequency
DAC_CAPSGL = 0x31   # DAC capture single
DAC_CAPCNT = 0x32   # DAC capture continuous

# TSC control
TSC_RPTVAL = 0x40

# Report data
RPT_FINISH = 0x80   # no more data to report
RPT_U08XXX = 0x81   # report with uint8_t data
RPT_S08XXX = 0x82   # report with int8_t data
RPT_U16XXX = 0x83   # report with uint16_t data
RPT_S16XXX = 0x84   # report with int16_t data
RPT_U32XXX = 0x85   # report with uint16_t data
RPT_S32XXX = 0x86   # report with int16_t data

# Command Code dictionary for packet decoding
CommandCodes = {
        SYS_SRESET: 'SYS_SRESET',
        SYS_WRESET: 'SYS_WRESET',
        SYS_STPMOD: 'SYS_STPMOD',
        SYS_DSLMOD: 'SYS_DSLMOD',
        SYS_SLPMOD: 'SYS_SLPMOD',
        DIO_SETVAL: 'DIO_SETVAL',
        DIO_RSTVAL: 'DIO_RSTVAL',
        DIO_GETVAL: 'DIO_GETVAL',
        DIO_TGLVAL: 'DIO_TGLVAL',
        ADC_SETMAG: 'ADC_SETMAG',
        ADC_SETFRQ: 'ADC_SETFRQ',
        ADC_CONSTV: 'ADC_CONSTV',
        ADC_SINEWV: 'ADC_SINEWV',
        ADC_SWTHWV: 'ADC_SWTHWV',
        ADC_TRNGWV: 'ADC_TRNGWV',
        DAC_SETFRQ: 'DAC_SETFRQ',
        DAC_CAPSGL: 'DAC_CAPSGL',
        DAC_CAPCNT: 'DAC_CAPCNT',
        RPT_FINISH: 'RPT_FINISH',
        RPT_U08XXX: 'RPT_U08XXX',
        RPT_S08XXX: 'RPT_S08XXX',
        RPT_U16XXX: 'RPT_U16XXX',
        RPT_S16XXX: 'RPT_S16XXX',
        RPT_U32XXX: 'RPT_U32XXX',
        RPT_S32XXX: 'RPT_S32XXX', 
        }


OutPackets = {
        'System Reset'      :bytes((PKT_HEADR,0x01,SYS_SRESET,
            SYS_SRESET)),
        'Watchdog Reset'    :bytes((PKT_HEADR,0x01,SYS_WRESET,
            SYS_WRESET)),
        'DIO 01 Set'        :bytes((PKT_HEADR,0x02,DIO_SETVAL,0x01,
            DIO_SETVAL ^ 0x01)),
        'DIO 01 Clear'      :bytes((PKT_HEADR,0x02,DIO_RSTVAL,0x01,
            DIO_RSTVAL ^ 0x01)),
        'DIO 01 Get'        :bytes((PKT_HEADR,0x02,DIO_GETVAL,0x01,
            DIO_GETVAL ^ 0x01)),
        'TSC Report Set'    :bytes((PKT_HEADR,0x02,TSC_RPTVAL,0x01,
            TSC_RPTVAL ^ 0x01)),
        'TSC Report Clear'  :bytes((PKT_HEADR,0x02,TSC_RPTVAL,0x00,
            TSC_RPTVAL ^ 0x00)),
        'Unknown'           :bytes((PKT_HEADR,0x05,0xFF,0x02,0x03,0x04,0x05,
            0xFF ^ 0x02 ^ 0x03 ^ 0x04 ^ 0x05)),
        }

################################################################################
#   This packet decoder should be called at the arrival of each data byte.
#   When valid packet arrives it returns data whose type is controlled by 
#   SetMode() function. Otherwise it returns None.
#
#   @code
#   # create an instance
#   self.pd = PacketDecoder()
#
#   # set mode
#   self.pd.SetMode('PAYLOAD')
#
#   # COM data event handler
#   def OnUpdateComData(self, evt):
#       # call decoder
#       ret = self.pd.AddByte(evt.byte[0])
#
#       # collect data
#       if ret is not None:
#           self.data.append(data)
#   @endcode
#
class PacketDecoder():

    def __init__(self, mode='FULL'):
        self.packet = bytearray()
        self.state = 'HDR'
        self.mode = mode
        self.len = 0

    ## Define the return data type
    #    
    # * 'FULL' returns full packet
    # * 'PAYLOAD' returns payload only
    # * 'DECODE' returns human readable string
    #
    def SetMode(self, mode):
        if mode.upper() in ('FULL', 'PAYLOAD', 'DECODE'):
            self.mode = mode.upper()

    ##Packet Decoding State Machine
    # Call this method every time new byte has arrived.
    def AddByte(self, byte):

        if self.state == 'HDR':
            # initialize the packet storage
            self.packet = bytearray()
            self.csum = 0

            # valid header received
            if byte == PKT_HEADR:
                self.packet.append(byte)
                # next state is LEN
                self.state = 'LEN'
                return None

            # ACK or NAK
            elif byte == PKT_ACK or byte == PKT_NAK or byte == PKT_IAM:
                # go back to HDR state
                self.state = 'HDR'

                if self.mode == 'FULL':
                    return byte

                elif self.mode == 'DECODE':
                    if byte == PKT_ACK:
                        return 'ACK'
                    elif byte == PKT_NAK:
                        return 'NAK'
                    elif byte == PKT_IAM:
                        return 'IAM'

                else:
                    return None

            # invalid byte
            else:
                # ignore the byte and go back to HDR state
                self.state = 'HDR'
                return None

        elif self.state == 'LEN':
            # invalid payload length
            if byte > MAX_PACKET:
                # return to HDR staet
                self.state = 'HDR'
                return None;

            # collect byte
            self.packet.append(byte)
            # save length
            self.len = byte
            # proceed to PLD state
            self.state = 'PLD'
            return None

        elif self.state == 'PLD':
            # collect byte
            self.packet.append(byte)
            # compute checksum
            self.csum = self.csum ^ byte

            # end of data
            if len(self.packet) >= 2 + self.len:
                # proceed to CSM state
                self.state = 'CSM'
            return None

        elif self.state == 'CSM':
            # checksum error
            if self.csum != byte:
                self.state = 'HDR'
                return None

            self.packet.append(byte)
            # start all over
            self.state = 'HDR'

            if self.mode == 'FULL' or self.mode == 'Full':
                return self.packet

            elif self.mode == 'PAYLOAD' or self.mode == 'Payload':
                return self.packet[2:-1]

            else:
                # command
                try:
                    txt = CommandCodes[self.packet[2]] + ':'
                except:
                    txt = 'Unknown :'

                # data
                txt = txt + self.packet[3:self.packet[1] + 2].hex()

                return txt


#--------1---------2---------3---------4---------5---------6---------7---------8
if __name__=='__main__':
    print('Unit Test for PacketDecoder()')

    # create an instance
    pd = PacketDecoder()

    # retrieve payload of the packet
    pd.SetMode('Payload')

    # feed the legit packets to the decoder
    for key, packet in OutPackets.items():

        for byte in packet:
            # feed byte to the decoder
            ret = pd.AddByte(byte)

            # no packet received
            if ret is None:
                pass
            elif ret == PKT_ACK:
                print('ACK received.')
            elif ret == PKT_NAK:
                print('NAK received.')
            elif ret == PKT_IAM:
                print('IAM received.')
            else:
                # return bytes should match with the payload of the packet
                if ret == packet[2:-1]:
                    print('packet received and match.')
                else:
                    print('packet received but does not match.')

    # set mode to decode
    pd.SetMode('decode')

    # feed the legit packets to the decoder
    for key, packet in OutPackets.items():

        for byte in packet:
            # feed byte to the decoder
            ret = pd.AddByte(byte)

            # no packet received
            if ret is None:
                pass
            else:
                print(ret)
